packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js

```diff
@@ -0,0 +1,158 @@
+const Identity = require('../../../../Identity');
+const IdentityPublicKeyDisabledAtWindowViolationError = require('../../../../../errors/consensus/state/identity/IdentityPublicKeyDisabledAtWindowViolationError');
+
+// security levels which must have at least one key
+const SECURITY_LEVELS = [
+  IdentityPublicKey.SECURITY_LEVELS.MASTER,
+];
+
+const BLOCK_TIME_WINDOW_MINUTES = 5;
+
+/**
+ * @param {StateRepository} stateRepository
+ * @param {validatePublicKeys} validatePublicKeys
+ * @param {validatePublicKeysAreEnabled} validatePublicKeysAreEnabled
+ * @return {validateIdentityUpdateTransitionState}
+ */
+function validateIdentityUpdateTransitionStateFactory(
+  stateRepository,
+  validatePublicKeys,
+  validatePublicKeysAreEnabled,
+) {
+  /**
+   * @typedef {validateIdentityUpdateTransitionState}
+   * @param {IdentityUpdateTransition} stateTransition
+   * @return {Promise<ValidationResult>}
+   */
+  // eslint-disable-next-line no-unused-vars
+  async function validateIdentityUpdateTransitionState(stateTransition) {
+    const result = new ValidationResult();
+
+    const identityId = stateTransition.getIdentityId();
+    const storedIdentity = await stateRepository.fetchIdentity(identityId);
+
+    // copy identity
+    const identity = (new Identity(storedIdentity.toObject()));
+
+    // Check revision
+    if (identity.getRevision() !== stateTransition.getRevision() - 1) {
+      result.addError(
+        new InvalidIdentityRevisionError(identityId.toBuffer(), identity.getRevision()),
+      );
+    }
+
+    const disablePublicKeys = stateTransition.getPublicKeyIdsToDisable();
+
+    if (disablePublicKeys) {
+      disablePublicKeys.forEach((id) => {
+        if (!identity.getPublicKeyById(id)) {
+          result.addError(
+            new InvalidIdentityPublicKeyIdError(id),
+          );
+        } else if (identity.getPublicKeyById(id).getReadOnly()) {
+          result.addError(
+            new IdentityPublicKeyIsReadOnlyError(id),
+          );
+        }
+      });
+
+      if (!result.isValid()) {
+        return result;
+      }
+
+      // Keys can only be disabled if another valid key is enabled in the same security level
+      disablePublicKeys.forEach(
+        (id) => identity.getPublicKeyById(id)
+          .setDisabledAt(stateTransition.getPublicKeysDisabledAt()),
+      );
+
+      const securityLevelsWithKeys = new Set();
+
+      SECURITY_LEVELS.forEach((securityLevel) => {
+        identity.getPublicKeys()
+          .filter((pk) => pk.getSecurityLevel() === securityLevel)
+          .forEach((pk) => {
+            if (!pk.getDisabledAt()) {
+              securityLevelsWithKeys.add(securityLevel);
+            }
+          });
+      });
+
+      SECURITY_LEVELS.forEach((securityLevel) => {
+        if (!securityLevelsWithKeys.has(securityLevel)) {
+          result.addError(new MissedSecurityLevelIdentityPublicKeyError(securityLevel));
```

Comment:
Shouldn't be a combination of purpose and security level? So we probably need to update error name and message and pass purpose there as well